messaging/messaging_test.go

```diff
@@ -0,0 +1,829 @@
+	},
+	{
+		name:   "InvalidTopicName",
+		tokens: []string{"token1"},
+		topic:  "foo*bar",
+		want:   "invalid topic name: \"foo*bar\"",
+	},
+	{
+		name:   "TooManyTokens",
+		tokens: strings.Split("a"+strings.Repeat(",a", 1000), ","),
+		topic:  "topic",
+		want:   "tokens list must not contain more than 1000 items",
+	},
+	{
+		name:   "EmptyToken",
+		tokens: []string{"foo", ""},
+		topic:  "topic",
+		want:   "tokens list must not contain empty strings",
+	},
+}
+
+func TestNoProjectID(t *testing.T) {
+	client, err := NewClient(context.Background(), &internal.MessagingConfig{})
+	if client != nil || err == nil {
+		t.Errorf("NewClient() = (%v, %v); want = (nil, error)", client, err)
+	}
+}
+
+func TestSend(t *testing.T) {
+	var tr *http.Request
+	var b []byte
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		tr = r
+		b, _ = ioutil.ReadAll(r.Body)
+		w.Header().Set("Content-Type", "application/json")
+		w.Write([]byte("{ \"name\":\"" + testMessageID + "\" }"))
+	}))
+	defer ts.Close()
+
+	ctx := context.Background()
+	client, err := NewClient(ctx, testMessagingConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client.fcmEndpoint = ts.URL
+
+	for _, tc := range validMessages {
+		t.Run(tc.name, func(t *testing.T) {
+			name, err := client.Send(ctx, tc.req)
+			if name != testMessageID || err != nil {
+				t.Errorf("Send() = (%q, %v); want = (%q, nil)", name, err, testMessageID)
+			}
+			checkFCMRequest(t, b, tr, tc.want, false)
+		})
+	}
+}
+
+func TestSendDryRun(t *testing.T) {
+	var tr *http.Request
+	var b []byte
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		tr = r
+		b, _ = ioutil.ReadAll(r.Body)
+		w.Header().Set("Content-Type", "application/json")
+		w.Write([]byte("{ \"name\":\"" + testMessageID + "\" }"))
+	}))
+	defer ts.Close()
+
+	ctx := context.Background()
+	client, err := NewClient(ctx, testMessagingConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client.fcmEndpoint = ts.URL
+
+	for _, tc := range validMessages {
+		t.Run(tc.name, func(t *testing.T) {
+			name, err := client.SendDryRun(ctx, tc.req)
+			if name != testMessageID || err != nil {
+				t.Errorf("SendDryRun() = (%q, %v); want = (%q, nil)", name, err, testMessageID)
+			}
+			checkFCMRequest(t, b, tr, tc.want, true)
+		})
+	}
+}
+
+func TestSendError(t *testing.T) {
+	var resp string
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Header().Set("Content-Type", "application/json")
+		w.Write([]byte(resp))
+	}))
+	defer ts.Close()
+
+	ctx := context.Background()
+	client, err := NewClient(ctx, testMessagingConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client.fcmEndpoint = ts.URL
+
+	cases := []struct {
+		resp string
+		want string
+	}{
+		{
+			resp: "{}",
+			want: "http error status: 500; reason: client encounterd an unknown error; response: {}",
```

Comment:
s/encounterd/encountered/, here and elsewhere

Reply:
Done